### ISSUE TITLE
[HL] Some weapon reloading animations run twice

### ISSUE THREAD

**Issue by Matthaiks** [ValveSoftware/halflife#2301]
Shoot and then spam the reload weapon key. The Glock, the MP5 and the Python will be reloaded twice. Python's second animation is cut, though.

This is not the same bug as https://github.com/ValveSoftware/halflife/issues/1076.

**Comment by rtxa** [comment]
I test it and I can confirm the issue

**Comment by SamVanheer** [comment]
This happens because weapon reloading was never fully implemented in the weapon prediction code on the client. As a result the client will think that the clip is not full and sets up another reload.

This is the code that changes the ammo values incorrectly:
https://github.com/ValveSoftware/halflife/blob/master/cl_dll/hl/hl_weapons.cpp#L329-L338

`iMaxClip()` returns 0 because item info is not cached like it is on the server.

I was able to fix this by changing the code to this:
```
#if 1 // FIXME, need ammo on client to make this work right
	// complete the reload. 
	ItemInfo ii;

	memset( &ii, 0, sizeof( ii ) );

	GetItemInfo( &ii );

	int j = V_min( ii.iMaxClip - m_iClip, m_pPlayer->m_rgAmmo[m_iPrimaryAmmoType]);

	// Add them to the clip
	m_iClip += j;
	m_pPlayer->m_rgAmmo[m_iPrimaryAmmoType] -= j;
#else	
	m_iClip += 10;
#endif
```

It's a bit ugly since it calls `GetItemInfo` every time but this will ensure it correctly adjusts the clip and spare ammo values during prediction.

A proper fix would require engine level changes to the data structures and networking system in order to network all ammo values which is impossible to do without breaking mods, so this will have to do.

Avoiding the `GetItemInfo` calls is possible by creating a list of all weapon classes on the client and caching the item info like the server does, but that's a lot of refactoring to do.

**Comment by Matthaiks** [comment]
cl_lw is set to "1" by default. "0" fixes this bug, it seems.

**Comment by SamVanheer** [comment]
Yes, because it disables the prediction code. But it's not a fix.

**Comment by sjain882** [comment]
> cl_lw is set to "1" by default. "0" fixes this bug, it seems.

That's a workaround not a fix
